src/solc_0.8/polygon/root/land/LandTunnelV2.sol

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.2;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import "../../../common/fx-portal/FxBaseRootTunnelUpgradeable.sol";
+import "../../../common/BaseWithStorage/ERC2771Handler.sol";
+import "../../../common/interfaces/ILandTokenV2.sol";
+import "../../../common/interfaces/IERC721MandatoryTokenReceiver.sol";
+
+/// @title LAND bridge on L1
+contract LandTunnelV2 is
+    FxBaseRootTunnelUpgradeable,
+    IERC721MandatoryTokenReceiver,
+    ERC2771Handler,
+    OwnableUpgradeable,
+    PausableUpgradeable
+{
+    ILandTokenV2 public rootToken;
+    bool internal transferringToL2;
+
+    event Deposit(address indexed user, uint256 size, uint256 x, uint256 y, bytes data);
+    event Withdraw(address indexed user, uint256 size, uint256 x, uint256 y, bytes data);
+
+    function initialize(
+        address _checkpointManager,
+        address _fxRoot,
+        ILandTokenV2 _rootToken,
+        address _trustedForwarder
+    ) public initializer {
+        rootToken = _rootToken;
+        __Ownable_init();
+        __Pausable_init();
+        __FxBaseRootTunnelUpgradeable_initialize(_checkpointManager, _fxRoot);
+        __ERC2771Handler_initialize(_trustedForwarder);
+    }
+
+    /// @dev called on ERC721 trasnfer to this contract
+    /// @param operator address of the one sending the ERC721 Token
+    /// @return onERC721Received function selector
+    function onERC721Received(
+        address operator,
+        address, /* from */
+        uint256, /* tokenId */
+        bytes calldata /* data */
+    ) external view override returns (bytes4) {
+        require(transferringToL2 || rootToken.isSuperOperator(operator), "LandTunnel: !BRIDGING");
+        return this.onERC721Received.selector;
+    }
+
+    /// @dev called on ERC721 batch trasnfer to this contract
+    /// @param operator address of the one sending the ERC721 Token
+    /// @return onERC721BatchReceived function selector
+    function onERC721BatchReceived(
+        address operator,
+        address, /* from */
+        uint256[] calldata, /* ids */
+        bytes calldata /* data */
+    ) external view override returns (bytes4) {
+        require(transferringToL2 || rootToken.isSuperOperator(operator), "LandTunnel: !BRIDGING");
+        return this.onERC721BatchReceived.selector;
+    }
+
+    function supportsInterface(bytes4 interfaceId) external pure returns (bool) {
+        return interfaceId == 0x5e8bf644 || interfaceId == 0x01ffc9a7;
+    }
+
+    function batchTransferQuadToL2(
+        address to,
+        uint256[] memory sizes,
+        uint256[] memory xs,
+        uint256[] memory ys,
+        bytes memory data
+    ) public whenNotPaused() {
+        require(to != address(0), "can't send to zero address");
+        require(sizes.length == xs.length && xs.length == ys.length, "l2: invalid data");
```

Comment:
Any particular reason for different formats for the errors (LandTunnel vs l2)?

Reply:
Used the error message of old landtunnel